### PR TITLE
Doctrine Inflector 2.0

### PR DESCRIPTION
**Pull request by jwage**
Refactor everything.

### CODE REVIEW

File: lib/Doctrine/Inflector/Inflect.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        $this->singularizer->reset();
+        $this->pluralizer->reset();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addRules(string $type, array $rules, bool $reset = false) : void
+    {
+        switch ($type) {
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Inflect.php]: Instead of having this, why don't we just have `addSingularRule(s)` and `addPluralRule(s)`?

File: lib/Doctrine/Inflector/Inflect.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+                return;
+
+            case 'singular':
+                $this->singularizer->addRules($rules, $reset);
+
+                return;
+        }
+
+        throw new InvalidArgumentException(sprintf('Invalid rule type specified "%s"', $type));
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Inflect.php]: This would need to be a package exception: `Doctrine\Inflector\Exception\InvalidType`.

File: lib/Doctrine/Inflector/Inflect.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        $this->singularizer = $singularizer;
+        $this->pluralizer   = $pluralizer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function tableize(string $word) : string
+    {
+        return strtolower(preg_replace('~(?<=\\w)([A-Z])~', '_$1', $word));
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Inflect.php]: This isn't unicode safe, needs to be mb_strtolower and use `u` regex modifier.

File: lib/Doctrine/Inflector/InflectInterface.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Inflector;
+
+interface InflectInterface
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/InflectInterface.php]: Should not be suffixed with `Interface` and class name imho shouldn't be a verb (what's wrong with `Inflector`?

File: lib/Doctrine/Inflector/InflectInterface.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+     *
+     * @return string The string with all delimiter-separated words capitalized.
+     */
+    public function ucwords(string $string, string $delimiters = " \n\t\r\0\x0B-") : string;
+
+    /**
+     * Clears Inflectors inflected value caches, and resets the inflection
+     * rules to the initial values.
+     */
+    public function reset() : void;
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/InflectInterface.php]: Should imho be removed. Resetting should be done by creating new instance. (Same for `add*` methods btw.)

File: lib/Doctrine/Inflector/InflectInterface.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * Adds custom inflection $rules, of either 'plural' or 'singular' $type.
+     *
+     * ### Usage:
+     *
+     * {{{
+     * Inflector::rules('plural', ['/^(inflect)or$/i' => '\1ables']);
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/InflectInterface.php]: Needs to be updated.

File: lib/Doctrine/Inflector/InflectorService.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Inflector;
+
+use function array_merge;
+use function is_array;
+
+abstract class InflectorService
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/InflectorService.php]: Not really convinced about having this thing here. :/

File: lib/Doctrine/Inflector/Pluralizer.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            return $word;
+        }
+
+        foreach ($this->rules['rules'] as $rule => $replacement) {
+            if (preg_match($rule, $word)) {
+                $this->cache['pluralize'][$word] = preg_replace($rule, $replacement, $word);
+
+                return $this->cache['pluralize'][$word];
+            }
+        }
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Pluralizer.php]: As Scrutinizer points out, this will implicitly return `null` which is incompatible with non-null return type.

File: lib/Doctrine/Inflector/Rules/Uninflected.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+namespace Doctrine\Inflector\Rules;
+
+class Uninflected
+{
+    /**
+     * Words that should not be inflected.
+     *
+     * @var string[]
+     */
+    private $uninflected = [
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Rules/Uninflected.php]: const?

File: lib/Doctrine/Inflector/InflectorService.php [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+{
+    /** @var mixed[] */
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/InflectorService.php]: Can't we do this better? Splitting into smaller units maybe?

File: lib/Doctrine/Inflector/InflectInterface.php (hunk 4 above)
Comment (jwage) [reply lib/Doctrine/Inflector/InflectInterface.php]: Right now I have left the old Inflector class with the static methods and have maintained BC with the old interface proxying to the new one. Once I have the tests moved over to the new API I will remove the existing Inflector class and rename things.

File: composer.json [hunk 11]
```diff
@@ -18,18 +18,19 @@
     "require-dev": {
         "doctrine/coding-standard": "^4.0",
         "phpstan/phpstan": "^0.9.2",
+        "phpstan/phpstan-phpunit": "^0.9.4",
```
Comment (carusogabriel) [inline composer.json]: I believe we need to import it to `phpstan.neon`. We have many `assertEquals`, and green build on Travis :(

Comment (jwage) [reply composer.json]: I changed all `self::assertEquals` calls to `self::assertSame`

File: lib/Doctrine/Inflector/Rules/Rules.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+    public function __construct(Rule ...$rules)
+    {
+        $this->rules = $rules;
+    }
+
+    public function inflect(string $word) : string
+    {
+        foreach ($this->rules as $rule) {
+            if (preg_match($rule->getFrom(), $word)) {
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Rules/Rules.php]: must use `!== 0`

File: lib/Doctrine/Inflector/Rules/Uninflected.php [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    public function isUninflected(string $word) : bool
+    {
+        return preg_match('/^(' . $this->getRegex() . ')$/i', $word, $regs) === 1;
+    }
+
+    private function getRegex() : string
+    {
+        if ($this->regex === null) {
+            $words = array_map(function (Word $word) {
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Rules/Uninflected.php]: return type missing

File: lib/Doctrine/Inflector/Rules/Uninflected.php [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+                return $word->getWord();
+            }, $this->words);
+
+            $this->regex = '(?:' . implode('|', $words) . ')';
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Rules/Uninflected.php]: Possible **regexp injection** here, each item in `words` must be properly escaped by `preg_quote()` with correct delimiter set as 2nd argument (possibly in the `array_map()` above).

File: lib/Doctrine/Inflector/RulesetInflector.php [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+    public function inflect(string $word) : string
+    {
+        $inflected = $this->ruleset->getIrregular()->inflect($word);
+
+        if ($inflected !== $word) {
+            return $inflected;
+        }
+
+        if ($this->ruleset->getUninflected()->isUninflected($word)) {
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/RulesetInflector.php]: Imho this check should be prioritized before attempting to inflect an _irregular_.

File: lib/Doctrine/Inflector/Rules/Ruleset.php [hunk 16]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Inflector\Rules;
+
+class Ruleset
+{
+    /** @var Rules */
+    private $rules;
```
Comment (Majkl578) [inline lib/Doctrine/Inflector/Rules/Ruleset.php]: Would it make sense to call this `$regular` (opposed to `$irregular`) instead?

File: lib/Doctrine/Inflector/Inflector.php [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    public function __construct(WordInflector $singularInflector, WordInflector $pluralInflector)
+    {
+        $this->singularInflector = $singularInflector;
+        $this->pluralInflector   = $pluralInflector;
+    }
+
+    /**
+     * Converts a word into the format for a Doctrine table name. Converts 'ModelName' to 'model_name'.
+     */
+    public function tableize(string $word) : string
```
[the remark this answers is not in the document]
Comment (jwage) [reply lib/Doctrine/Inflector/Inflector.php]: I am not sure honestly. The definition of inflection is broad enough to include the tableize, classify, etc. methods. If you follow that logic, you'd have a different class for each method. That is a pain for the end user to have to have multiple classes/services.

File: lib/Doctrine/Inflector/InflectorFactory.php [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+namespace Doctrine\Inflector;
+
+use Doctrine\Inflector\Rules\Irregular;
+use Doctrine\Inflector\Rules\Plural;
+use Doctrine\Inflector\Rules\Rules;
+use Doctrine\Inflector\Rules\Ruleset;
+use Doctrine\Inflector\Rules\Singular;
+use Doctrine\Inflector\Rules\Uninflected;
+
+class InflectorFactory
```
[the remark this answers is not in the document]
Comment (jwage) [reply lib/Doctrine/Inflector/InflectorFactory.php]: Issue recorded here https://github.com/doctrine/inflector/issues/94

File: lib/Doctrine/Inflector/Rules/Irregular.php [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    /** @var Rule[] */
+    private $rules;
+
+    /** @var string|null */
+    private $regex;
+
+    public function __construct(Rule ...$rules)
+    {
+        foreach ($rules as $rule) {
+            $this->rules[$rule->getFrom()] = $rule;
```
[the remark this answers is not in the document]
Comment (jwage) [reply lib/Doctrine/Inflector/Rules/Irregular.php]: Issue recorded here https://github.com/doctrine/inflector/issues/93

**Comment by teohhanhui** [comment]
What's the upgrade path from 1.x to 2.0? Will there be a 1.x version which will help users migrate to 2.0?

**Comment by jwage** [comment]
@teohhanhui The upgrade path is like this:

    $plural = Inflector::pluralize('base');

Would be changed to this:

    $inflector = new Inflector(
        new CachedWordInflector(new RulesetInflector(
            English\Rules::getSingularRuleset()
        )),
        new CachedWordInflector(new RulesetInflector(
            English\Rules::getPluralRuleset()
        ))
    );
    $plural = $inflector->pluralize('base');

So basically we're just not using static methods anymore and you have to instantiate the Inflector. You can see the 2.0 docs here https://www.doctrine-project.org/projects/doctrine-inflector/en/2.0/

**Comment by garrettw** [comment]
@jwage relax, I'm just messing with you 😉
